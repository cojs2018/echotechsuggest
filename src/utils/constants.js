export const bookmarksAPI = 'https://5zcbdoz0h3.execute-api.eu-west-2.amazonaws.com/dev';
export const tagsAPI = 'https://141b51xkk2.execute-api.eu-west-2.amazonaws.com/dev';