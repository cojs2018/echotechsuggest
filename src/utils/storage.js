import { writeFile } from 'fs';

export async function createBookmark(bookmarkUrl) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');

        let bookmarkLocation = bookmarkUrl;
        if(!bookmarkLocation.includes('http')) {
            bookmarkLocation = `http://${bookmarkLocation}`;
        }

        const newUrlData = new URL(bookmarkLocation);

        const bookmarkJSON = {
            path: bookmarkUrl,
            url: newUrlData.toJSON(),
            createdAt: Date.now()
        }
        const articleName = `${newUrlData.pathname} - ${newUrlData.hostname}`;

        bookmarkUrlList[articleName] = bookmarkJSON;

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark created successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not create bookmark!' 
        }));
    }
}

export async function listBookmarks() {
    const bookmarkUrlList = require('../../bookmarkUrls.json');
    return bookmarkUrlList;
}

export async function deleteBookmark(bookmarkName) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');
        
        if(!bookmarkUrlList[bookmarkName]) {
            return {
                status: 200,
            }
        }

        delete bookmarkUrlList[bookmarkName]

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark deleted successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not delete bookmark!' 
        }));
    }
}