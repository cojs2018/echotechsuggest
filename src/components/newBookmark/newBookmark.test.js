import React from 'react';
import render from 'react-test-renderer';
import { mount } from 'enzyme';
import NewBookmark from './newBookmark';

jest.mock('../../utils/storage')
import { createBookmark } from '../../utils/storage';

describe('New Bookmark', () => {
    it('renders correctly', () => {
        const tree = render
            .create(<NewBookmark />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    })
})

test('Add url and submit, no error', async () => {
    let bookmarkUrl = '';
    let message = '';
    let error = false;

    const setBookmarkUrl = jest.fn(async (newUrl) => bookmarkUrl = newUrl);
    const setMessage = jest.fn(async (newMessage) => message = newMessage);
    const setError = jest.fn(async (bool) => error = bool);

    React.useState = jest.fn()
        .mockReturnValueOnce([bookmarkUrl, setBookmarkUrl])
        .mockReturnValueOnce([message, setMessage])
        .mockReturnValueOnce([error, setError]);

    createBookmark.mockImplementationOnce((bookmarkURL) => new Promise((resolve, reject) => {
        resolve({
            status: 200,
            message: 'Bookmark created successfully!'
        })
    }))

    const wrapper = mount(<NewBookmark />);

    const urlInput = wrapper.find('#url').find('input').at(0);
    
    const eventPayload = { target: { value: 'www.example.com' } };
    urlInput.simulate('change', eventPayload);

    await new Promise(resolve => setImmediate(resolve))
    expect(bookmarkUrl).toStrictEqual(eventPayload.target.value);

    const submitButton = wrapper.find('#submit').find('button').at(0);
    submitButton.simulate('click');

    await new Promise(resolve => setImmediate(resolve))
    expect(message).toStrictEqual('Url has now been stored, please await your results.');
    expect(error).toBeFalsy();
});

test('error', async () => {
    let bookmarkUrl = '';
    let message = '';
    let error = false;

    const setBookmarkUrl = jest.fn(async (newUrl) => bookmarkUrl = newUrl);
    const setMessage = jest.fn(async (newMessage) => message = newMessage);
    const setError = jest.fn(async (bool) => error = bool);

    React.useState = jest.fn()
        .mockReturnValueOnce([bookmarkUrl, setBookmarkUrl])
        .mockReturnValueOnce([message, setMessage])
        .mockReturnValueOnce([error, setError]);

    createBookmark.mockImplementationOnce((bookmarkURL) => new Promise((resolve, reject) => {
        reject({
            status: 500,
            message: 'Not a url!'
        })
    }))

    const wrapper = mount(<NewBookmark />);

    const urlInput = wrapper.find('#url').find('input').at(0);
    
    const eventPayload = { target: { value: 'wwwexamplecom' } };
    urlInput.simulate('change', eventPayload);

    await new Promise(resolve => setImmediate(resolve))
    expect(bookmarkUrl).toStrictEqual(eventPayload.target.value);

    const submitButton = wrapper.find('#submit').find('button').at(0);
    submitButton.simulate('click');

    await new Promise(resolve => setImmediate(resolve))
    expect(message).toBeDefined();
    expect(error).toBeTruthy();
})