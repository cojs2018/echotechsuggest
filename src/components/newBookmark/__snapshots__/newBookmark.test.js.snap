// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Bookmark renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "maxHeight": "50%",
      "maxWidth": "97%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
        },
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 16,
              "letterSpacing": 0.5,
              "lineHeight": 24,
              "marginVertical": 2,
            },
            undefined,
          ],
        ],
      ]
    }
  >
    New Bookmark
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.12)",
          "height": 0.5,
        },
        undefined,
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "maxWidth": "90%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.26)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
          }
        }
      />
      <View
        style={
          Object {
            "minHeight": 64,
            "paddingBottom": 0,
            "paddingTop": 0,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Add url to article you wish to bookmark"
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "zIndex": 1,
              },
              Object {
                "paddingLeft": 12,
                "paddingRight": 12,
              },
              Object {
                "height": 64,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              Array [
                Object {},
              ],
            ]
          }
          testID="url"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#6200ee",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
        testID="send"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
